pygdf/libgdf_groupby.py

```diff
@@ -0,0 +1,246 @@
+# Copyright (c) 2018, NVIDIA CORPORATION.
+
+import numpy as np
+
+from numba import cuda
+
+from .dataframe import DataFrame, Series
+from .buffer import Buffer
+
+from libgdf_cffi import ffi, libgdf, GDFError
+
+
+# import pytest
+
+print("class")
+
+
+class LibGdfGroupby(object):
+    """Groupby object returned by pygdf.DataFrame.groupby().
+    """
+
+    def __init__(self, df, by, method="GDF_SORT"):
+        """
+        Parameters
+        ----------
+        df : DataFrame
+        by : str of list of str
+            Column(s) that grouping is based on.
+            It can be a single or list of column names.
+        """
+        print("__init__")
+
+        self._df = df
+        self._by = [by] if isinstance(by, str) else list(by)
+        self._val_columns = [idx for idx in self._df.columns
+                             if idx not in self._by]
+        if (method == "GDF_SORT"):
+            self._method = libgdf.GDF_SORT
+        else:
+            self._method = libgdf.GDF_HASH
+
+    def _apply_agg(self, agg_type, result, add_col_values,
+                   ctx, val_columns, val_columns_out=None):
+
+        if (val_columns_out is None):
+            val_columns_out = val_columns
+
+        ncols = len(self._by)
+        cols = [self._df[thisBy]._column.cffi_view for thisBy in self._by]
+
+        first_run = add_col_values
+        need_to_index = len(
+            self._val_columns) > 0 and self._method == libgdf.GDF_HASH
+
+        col_count = 0
+        for val_col in val_columns:
+            col_agg = self._df[val_col]._column.cffi_view
+
+# assuming here that if there are multiple aggregations that the
+# aggregated results will be in the same order for GDF_SORT method
+            if need_to_index:
+                out_col_indices_series = Series(
+                    Buffer(cuda.device_array(col_agg.size, dtype=np.int32)))
+                out_col_indices = out_col_indices_series._column.cffi_view
+            else:
+                out_col_indices = ffi.NULL
+
+            if first_run:
+                out_col_values_series = [Series(Buffer(cuda.device_array(
+                    col_agg.size, dtype=self._df[self._by[i]]._column.data.dtype))) for i in range(0, ncols)]
+                out_col_values = [
+                    out_col_values_series[i]._column.cffi_view for i in range(0, ncols)]
+            else:
+                out_col_values = ffi.NULL
+
+            if agg_type == "count":
+                out_col_agg_series = Series(
+                    Buffer(cuda.device_array(col_agg.size, dtype=np.int64)))
+            else:
+                out_col_agg_series = Series(Buffer(cuda.device_array(
+                    col_agg.size, dtype=self._df[val_col]._column.data.dtype)))
+
+            out_col_agg = out_col_agg_series._column.cffi_view
+
+            if agg_type == "min":
```

Comment:
It seems as though there should be an `agg_funcs` dict on th class which maps the names to the functions and then this is:

```python
agg_func = self.agg_funcs.get(agg_type, None)
if agg_func is None:
    raise RuntimeError("ERROR: this aggregator has not been implemented yet")
err = agg_func(ncols, cols, col_agg, out_col_indices, out_col_values, out_col_agg, ctx)
```

rather than this big if-elif block.